arch/x86/core/elfloader_x86.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+#include <elfloader.h>
+
+#ifdef CONFIG_64BIT
+
+/* x86 64-bit relocation for DYN file */
+void elfloader_arch_relocate_dyn(zmodule_t *module, elf_rel_t *rel,
+		elf_addr sym_addr)
+{
+}
+
+/* x86 64-bit relocation for REL file */
+void elfloader_arch_relocate_rel(elf_rel_t *rel)
+{
+}
+
+#else
+
+/* x86 32-bit relocation for DYN file */
+void elfloader_arch_relocate_dyn(zmodule_t *module, elf_rel_t *rel,
+		elf_addr sym_addr)
+{
+	elf_addr *where;
+
+	where = (elf_addr *)(module->load_start_addr
+			+ rel->r_offset - module->virt_start_addr);
+	switch (ELF32_R_TYPE(rel->r_info)) {
+	case R_386_GLOB_DAT:
+	case R_386_JUMP_SLOT:
+		*where = sym_addr;
+		break;
+	case R_386_RELATIVE:
+		*where = sym_addr + *where;
+		break;
+	default:
+		/* unsupported relocation type */
+		break;
```

Comment:
FWIW: this complexity is going to hurt us badly at the arch level, where lots of architectures have a LOT of very weird relocation formats.

I've made the point in the past, but it wouldn't actually be difficult to go to a "GOT-only" relocation strategy, where the only relocations we support are to the GOT table emitted by compiler-generated ELF PIC code.  Then this code would become almost entirely portable (I'm sure there are edge cases).

But basically: having to support every relocation type for every architecture seems sisyphean to me.

Reply:
elfloader_arch_relocate_dyn this is used to handle GOT-only relocation, the input EFL file is PIC code, but for the relocation type, I only listed those common types.

for the elfloader_arch_relocate_rel, I want to use this for loading an executable file and treat it as a relocatable file, but seems that I am wrong, EXE should also be generated using PIE, it's relocations should also using "GOT"